src/hook/pauseResume.js

```diff
@@ -0,0 +1,299 @@
+			return currValue + Number(match) * sign;
+		};
+	}
+
+	AniProperty.prototype.init = function() {
+		this.start = now();
+		this.elapsed = 0;
+	};
+
+	AniProperty.prototype.addEasingFn = function(easingName) {
+		this.easingNames.push(easingName);
+	};
+
+	AniProperty.prototype.clearEasingFn = function() {
+		var easing;
+		while (easing = this.easingNames.shift()) {
+			delete $.easing[easing];
+		}
+		this.easingNames = [];
+	};
+
+	function addAniProperty(el, prop, optall) {
+		var prevProp;
+		var propCount = el.__aniProps ? el.__aniProps.length : 0;
+		var isFirst = false;
+		var newProp;
+
+		if (propCount === 0) {
+			prevProp = el.style;
+			isFirst = true;
+		} else {
+			prevProp = el.__aniProps[propCount - 1].prop;
+		}
+
+		//prevProp is used for calculating absolute value by accumulating aniProps.
+		//So newProp has absolute value accumlated.
+		newProp = new AniProperty(prop, optall, prevProp, isFirst);
+		el.__aniProps = el.__aniProps || [];
+
+		//Animation is excuted immediately.
+		if (el.__aniProps.length === 0) {
+			newProp.init();
+		}
+		el.__aniProps.push(newProp);
+	}
+
+	function getOptAll(speed, easing, fn) {
+		var opt = speed && typeof speed === "object" ?
+		jQuery.extend({}, speed) : {
+			complete: fn || !fn && easing ||
+				jQuery.isFunction(speed) && speed,
+			duration: speed,
+			easing: fn && easing ||
+					easing && !jQuery.isFunction(easing) && easing
+		};
+
+		opt.duration = jQuery.fx.off ? 0 : typeof opt.duration === "number" ?
+			opt.duration : opt.duration in jQuery.fx.speeds ?
+				jQuery.fx.speeds[opt.duration] : jQuery.fx.speeds._default;
+
+		// normalize opt.queue - true/undefined/null -> "fx"
+		if (opt.queue == null || opt.queue === true) {
+			opt.queue = "fx";
+		}
+
+		return opt;
+	}
+
+	$.fn.animate = function(prop, speed, easing, callback) {
+		return this.each(function() {
+			//optall should be made for each elements.
+			var optall = getOptAll(speed, easing, callback);
+			var orginalComplete = optall.complete;
+
+			//Override to check current animation is done.
+			optall.complete = function() {
+				//Dequeue animation property that was ended.
+				var removeProp = this.__aniProps.shift();
+				removeProp.clearEasingFn();
+
+				// console.log("ani complete, shift prop:", removeProp);
+				this.__aniProps[0] && this.__aniProps[0].init();
+				if (orginalComplete && typeof(orginalComplete) === "function") {
+					orginalComplete.call(this);
+				}
+			};
+
+			//Queue animation property to recover the current animation.
+			addAniProperty(this, prop, optall);
+			animateFn.call($(this), prop, optall);
+		});
+
+		// TODO: Below code is more reasonable?
+		// return animateFn.call(this, prop, optall); // and declare optall at outside this.each loop.
+	};
+
+	$.fn.pause = function() {
+		// console.warn("===========PAUSE==============");
+		return this.each(function() {
+			var p;
+			var type = "fx";
+
+			if (!this.__aniProps) {
```

Comment:
@mixed This condition is somewhat different. so I didn't make it common.

first one check if
```js
this.__aniProps[0].paused === true
```
and last one check if,
```js
this.__aniProps[0].paused === false
```

But I agree it has some repeated code, and I'll consider more good solution.